### ISSUE TITLE
Add instructions for integrating ESM version with snowpack

### ISSUE THREAD

**Issue by jesperp** [microsoft/monaco-editor#2227]
Can we have instructions on how to integrate the ESM version of Monaco with Snowpack?
As in here:
https://github.com/microsoft/monaco-editor/blob/master/docs/integrate-esm.md

I'm currently trying different configurations and all seems to work except when workers are created. If any one has a simple setup to share I can try it and have a PR to update the docs

**Comment by Prinzhorn** [comment]
> except when workers are created

Do you by any chance get an "Unexpected usage" error? From what I understand this might be broken in ESM https://github.com/microsoft/monaco-editor/issues/2212

**Comment by jesperp** [comment]
Yes! Basically I have this (unnecessary stuff left out for readability)

Main application typescript:
```javascript
import * as monaco from 'monaco-editor/esm/vs/editor/editor.api.js';
import 'monaco-editor/esm/vs/editor/editor.worker.js'
import 'monaco-editor/esm/vs/language/typescript/ts.worker.js'
import 'monaco-editor/esm/vs/editor/browser/controller/coreCommands.js';
import 'monaco-editor/esm/vs/editor/contrib/find/findController.js';
import 'monaco-editor/esm/vs/language/typescript/monaco.contribution.js';
import 'monaco-editor/esm/vs/basic-languages/typescript/typescript.contribution.js';

self.MonacoEnvironment = {
	getWorkerUrl: function (moduleId, label) {
		if (label === 'typescript' || label === 'javascript') {
			return 'web_modules/monaco-editor/esm/vs/language/typescript/ts.worker.js';
		}
		return 'web_modules/monaco-editor/esm/vs/editor/editor.worker.js';
	}
}
```

By just doing an import without using, Snowpack makes the module available under `http://localhost/web_modules/monaco-editor/...`. Monaco does load the file successfully with a 200. But then I get a `SyntaxError: import declarations may only appear at top level of a module` for the first line of `ts.worker.js` which is a sign that it is not loaded as a ESM module.

Not sure wether to solve this in my JS-file, Snowpack or Monaco..
Thanks for your suggestion, I will check your issue!

**Comment by jesperp** [comment]
Steps to reproduce:

```
npx create-snowpack-app monaco-test --template @snowpack/app-template-blank-typescript --use-yarn
cd monaco-test
yarn add monaco-editor
```

edit: **src/index.ts**
```typescript
import * as monaco from 'monaco-editor/esm/vs/editor/editor.api.js'
import 'monaco-editor/esm/vs/editor/editor.worker.js'
import 'monaco-editor/esm/vs/language/typescript/ts.worker.js'
import 'monaco-editor/esm/vs/language/typescript/monaco.contribution.js';
import 'monaco-editor/esm/vs/basic-languages/typescript/typescript.contribution.js';


(self as any).MonacoEnvironment = {
    getWorker: function(moduleId:string, label:string) {
    console.log("Getting worker", moduleId, label)
    if (['javascript', 'typescript'].includes(label)) {
      return new Worker('web_modules/monaco-editor/esm/vs/language/typescript/ts.worker.js', {
        type: 'module'
      })
    }
    else {
      return new Worker('web_modules/monaco-editor/esm/vs/editor/editor.worker.js', {
        type: 'module'
      })
    }
  }
}


const app = document.getElementById('app')
if (app !== null) {
  monaco.editor.create(app, {
    value: 'function x():void {console.log("Hello world!"); }',
    language: 'typescript'
  })
}

```

edit: **public/index.html**
```html
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8" />
    <title>Snowpack App</title>
  </head>
  <body>
    <noscript>You need to enable JavaScript to run this app.</noscript>
    <script type="module" src="/_dist_/index.js"></script>
    <main id="app" style="width:800px; height:600px;"></main>
  </body>
</html>

```

And then run with `yarn start` (this can take a while first time).

**Comment by naknode** [comment]
@jesperp I am doing the same exact thing so let us know if you found a solution.

**Comment by alexdima** [comment]
I have just tried the sample from https://github.com/microsoft/monaco-editor/issues/2227#issuecomment-731578362 and it works for me (I just added one more import to `'monaco-editor/esm/vs/editor/editor.all.js'`). There seems to be a problem with the font for icons, but things appear to work:

![image](https://user-images.githubusercontent.com/5047891/99903790-f6c22f00-2cc6-11eb-8af2-65e036dc4e0d.png)

**Comment by jesperp** [comment]
@alexdima Thanks for your comment! Could you add your full working example? I tried adding just `import 'monaco-editor/esm/vs/editor/editor.all.js'` to my index.ts file but doesn't seem to work. Same error as before

**editor.worker.js:1**
`SyntaxError: import declarations may only appear at top level of a module`

**Comment by jesperp** [comment]
Hmm, so this may be a browser issue. I tried my initial example and alexdima's with the extra import in Chrome 86. It works as expected.

### Firefox 83.0 fails:
![Screenshot 2020-11-22 at 15 51 54](https://user-images.githubusercontent.com/376921/99907018-b5d41580-2cda-11eb-9ce0-3c7e9aede968.png)



### Safari 14.0.1 fails:
![Screenshot 2020-11-22 at 15 51 28](https://user-images.githubusercontent.com/376921/99907022-be2c5080-2cda-11eb-8fe3-e98fe7c138b2.png)

**Comment by wires** [comment]
I am struggling to get this to work well with a snowpack + Svelte application. Been trying the past three days but keep running into worker related issues. I'm not even able to get the AMD version (`/package/min`) working from a `/vendor` mount (didn't keep a good debugging log, so can't be specific rn).

The furthest I got was based off this working gist, but I have not been able to extend that config to support Svelte.
https://gist.github.com/mattpowell/221f7d35c4ae1273dc2e1ee469d000a7 

HTH

**Comment by wires** [comment]
I gave it another try, starting from `create-snowpack-app`. It managed to get it to work for `snowpack dev` but not `snowpack build`.

The `web_modules` folder is not populated with the modules from `install`.

See the readme here for more details https://github.com/wires/debugging-snowpack-svelte-monaco

**Comment by wires** [comment]
Ha, okay, managed to get this to work, the problem was I wasn't referencing the modules.

See the updated readmed here for details https://github.com/wires/debugging-snowpack-svelte-monaco

**Comment by alexrp** [comment]
@wires I don't suppose you've managed to get this working with Snowpack 3.x?

**Comment by LumaKernel** [comment]
Here is my attempt to work with Snowpack 3.x. https://github.com/luma-dev/alt-json

It bundles worker js files into single file before the snowpack compilation with [esbuild](https://github.com/evanw/esbuild). It really compiles fast so you wouldn't feel any stresses in development. Even if using esbuild, Snowpack is still necessary to treat css and font assets files imported in `monaco-editor`.

Main point is that Snowpack v3.x is exposing local `node_modules/` to `//.../node_modules/...` instead of `//.../web_modules/...`. But even so, Snowpack prepends the `import ...` statements for controlling and do not bundle, so it does not work with monaco Web Worker system. We  cannot use ESModule resolution in the worker context and should prepare worker js files in one file.

**Comment by remcohaszing** [comment]
I think the example in #2605 solves this issue.
